### ISSUE TITLE
Universal loading with React and Redux

### ISSUE THREAD

**Issue by bkbrooks** [i18next/react-i18next#38]
Hello,

I'm trying to create an application with localization that also uses universal loading. I feel like I'm almost there but I'm having an issue with loading the user's language. 

I am using 2 other modules to help me with that: 
[https://github.com/i18next/i18next-node-fs-backend](https://github.com/i18next/i18next-node-fs-backend) (instead of i18next-xhr-backend in your example)
[https://github.com/i18next/i18next-express-middleware](https://github.com/i18next/i18next-express-middleware)

Here is what my i18n init looks like:

```
import i18n from 'i18next'
import Backend from 'i18next-node-fs-backend'
import { LanguageDetector } from 'i18next-express-middleware'

i18n
  .use(Backend)
  .use(LanguageDetector)
  .init({
    whitelist: ['en','en-US','fr'],
    fallbackLng: 'en',

    // have a common namespace used around the full app
    ns: ['common'],
    defaultNS: 'common',

    debug: true,

    interpolation: {
      escapeValue: false // not needed for react!!
    },

    backend: {
      // path where resources get loaded from
      loadPath: 'locales/{{lng}}/{{ns}}.json',

      // path to post missing resources
      addPath: 'locales/{{lng}}/{{ns}}.missing.json',

      // jsonIndent to use when storing json files
      jsonIndent: 2
    }
  })

export default i18n
```

Then in my express server, I am taking i18n and using it with express so that I can detect the user's language:

```
import i18n from './i18n'
import { I18nextProvider } from 'react-i18next'
var i18nMiddleware = require('i18next-express-middleware')

const app = express()
app.use(i18nMiddleware.handle(i18n))

app.get('*', (req, res) => {

  console.log('language ' + req.language)
  console.log('i18n language ' + req.i18n.language)
  console.log(req.i18n.t('app.appName'))

  match({ routes, location: req.url }, (err, redirect, props) => {
      const contentHtml = ReactServer.renderToString(
        <I18nextProvider i18n={ req.i18n }>
          <Provider store={store} key="provider">
            <RouterContext {...props} />
          </Provider>
        </I18nextProvider>
      )
      res.send(layout(contentHtml, store.getState()))
  })
})
```

I am able to grab `req.language` and `req.i18n.language`. Also, `req.i18n.t()` works for the keys I pass into it. The problem is when I try and pass in `req.i18n` into the I18nextProvider, it throws an error saying `i18n.loadNamespaces is not a function`. 

Do you have any other suggestions about how to use this? Or maybe there's an attribute I can add to I18nextProvider for the language to use for the user? Maybe something like this:

```
<I18nextProvider i18n={ i18n } language={req.language}>
```

Or maybe I'm going about this all wrong and you have another idea for Universally loading a React/Redux application with i18next capabilities.

Thanks for your help!

**Comment by jamuhl** [comment]
the point req.i18n is not a 'real' i18next instance...it's only a facade to the most used functions needed in an req flow.

try passing a clone of i18n you imported from i18next to the provider. http://i18next.com/docs/api/#clone-instance --> set the language on clone to the req.language

while on client the translated hoc triggers a rerender on loaded defined namespace on server you need to make sure that the namespaces and languages are preloaded for server side rendering.

not yet done universal loading...so i'm happy to see one try.

**Comment by bkbrooks** [comment]
Thanks! This worked:

```
let useri18n = i18n.cloneInstance()
useri18n.changeLanguage(req.language)

...

<I18nextProvider i18n={ useri18n }>
  <Provider store={store} key="provider">
    <RouterContext {...props} />
  </Provider>
</I18nextProvider>
```

I can get the client side set up now, but I'm curious about your statement:

> you need to make sure that the namespaces and languages are preloaded for server side rendering

How do you suggest doing that?

**Comment by jamuhl** [comment]
basically by adding all namespaces used in client to the ns: ['common'] on server, plus adding preload: ['en','en-US','fr'] to init options

**Comment by bkbrooks** [comment]
Cool, thanks! I kind of figured that's what you were meaning. 

And I assume that on the client side, you'd have to load the namespaces again asynchronously by using the i18next-xhr-backend module. It seems redundant to do both, but I think that's just how this Universal thing is supposed to work.

**Comment by jamuhl** [comment]
you might add the needed resources to some global object on the prerendered page and add that as resources to i18next.init on client...but honestly no win in compare to loading it via xhr-backend...the resources are transfered over the wire anyway.

so yes...using xhr-backend and load it again...seems to be it.

if you got all up i would love to here from you again...would somehow be great to have some 'how to' or even sample here for others to follow.

**Comment by loudwinston** [comment]
I have a fork that adds support for universal loading.  It works similarly to the `loadOnServer` method of https://www.npmjs.com/package/redux-async-connect

I will put up a PR soon.

**Comment by jamuhl** [comment]
@loudwinston that's awesome!!!

**Comment by loudwinston** [comment]
@jamuhl Not sure about the timeline for this, I have to get it approved by the risk/security team at my company before I can submit a PR.  I don't know exactly how long that's going to take.  

However I can tell you the basic approach - 

I modified the `@translate` decorator to write a `namespaces` static property on the returned class.  I then wrote a `loadNamespaces` method that accepts a list of components (usually generated by the `react-router` `match` function) and an i18n instance.  

The `loadNamespaces` method recurses through all of the components passed to it and looks for the `namespaces` static property, then calls the `i18n.loadNamespaces` method on the `i18n` instance that was passed in with all of these namespaces.  It returns a promise that will resolve once `i18n` has completed loading all of those namespaces.  You can then wait for this promise to resolve before rendering your React views.

This is basically the same technique that `redux-async-connect` uses to preload data for server-side rendering.  

Let me know if this sounds like a reasonable approach to you.  I'm hoping to get a PR up within the week.

**Comment by jamuhl** [comment]
sounds great...risk/security team...sounds like a big company...we don't have such teams ;)

**Comment by loudwinston** [comment]
Fairly large company yes.  I probably should have just made the changes on my own time, then I wouldn't have to deal with this.  Although then I wouldn't have gotten paid for it, so... :)

**Comment by simpleblack** [comment]
+1

**Comment by simpleblack** [comment]
Hello. I used universal rendering. 
This project https://github.com/erikras/react-redux-universal-hot-example
I did add react-i18next.
This my repo https://github.com/simpleblack/react-redux-universal-hot-example

I had client error issue.
Error message is
```Warning: Failed Context Types: Required context `i18n` was not specified in `Translate[Home]`. Check the render method of `RouterContext`.```
```Uncaught TypeError: Cannot read property 'loadNamespaces' of undefined```

``` javascript
<I18nextProvider i18n={useri18n}>
  <Provider store={store} key="provider">
    <ReduxAsyncConnect {...renderProps} />
  </Provider>
</I18nextProvider>
```

I didn't solve the problem.
Thanks for your help!

**Comment by jamuhl** [comment]
sounds like i18n instance is not passed down via context....not sure where it happens...sorry

**Comment by sorahn** [comment]
Hmm, I ran into something similar and I thought it might have been an enzyme bug. I'm using enzyme with mocha and jsdom, and I'm getting a TypeError as well.

`TypeError: Cannot read property 'loadNamespaces' of undefined`

Check out the code here: https://github.com/sorahn/enzyme-bug

**Comment by loudwinston** [comment]
@jamuhl I've just put up a PR with support for universal rendering.  https://github.com/i18next/react-i18next/pull/52

Let me know if you have any questions about it, I'm happy to discuss :)

**Comment by simpleblack** [comment]
Thanks. @jamuhl @sorahn 
This worked. 

I added react-i18next for universal rendering.
This example repo.
https://github.com/simpleblack/react-redux-universal-hot-example

**Comment by jamuhl** [comment]
@loudwinston @sorahn @simpleblack @bkbrooks thank you all...i really appreciate your efforts put into this...great to see this working

**Comment by jamuhl** [comment]
published react-i18next@1.3.0 adding @loudwinston https://github.com/i18next/react-i18next/pull/52

**Comment by loudwinston** [comment]
@jamuhl Thanks!  Glad to help out.
